src/main/java/ch/njol/skript/variables/FlatFileStorage.java

```diff
@@ -18,180 +18,259 @@
+						// Couldn't deserialize variable
 						if (invalid.length() != 0)
 							invalid.append(", ");
+
 						invalid.append(split[0]);
-						unsuccessful++;
+						unsuccessfulVariableCount++;
 						continue;
 					}
-					if (d instanceof String && update2_0_beta3) {
-						d = Utils.replaceChatStyles((String) d);
+
+					// Legacy
+					if (deserializedValue instanceof String && update2_0_beta3) {
+						deserializedValue = Utils.replaceChatStyles((String) deserializedValue);
 					}
-					Variables.variableLoaded("" + split[0], d, this);
+
+					Variables.variableLoaded(split[0], deserializedValue, this);
 				}
 			}
-		} catch (final IOException e) {
+		} catch (IOException e) {
 			loadError = true;
-			ioEx = e;
-		} finally {
-			if (r != null) {
-				try {
-					r.close();
-				} catch (final IOException e) {}
-			}
+			ioException = e;
 		}
-		
-		final File file = this.file;
-		if (file == null) {
-			assert false : this;
-			return false;
-		}
-		
-		if (ioEx != null || unsuccessful > 0 || update2_1) {
-			if (unsuccessful > 0) {
-				Skript.error(unsuccessful + " variable" + (unsuccessful == 1 ? "" : "s") + " could not be loaded!");
+
+		if (ioException != null || unsuccessfulVariableCount > 0 || update2_1) {
+			// Something's wrong (or just an old version)
+			if (unsuccessfulVariableCount > 0) {
+				Skript.error(unsuccessfulVariableCount + " variable" + (unsuccessfulVariableCount == 1 ? "" : "s") + " could not be loaded!");
 				Skript.error("Affected variables: " + invalid.toString());
 			}
-			if (ioEx != null) {
-				Skript.error("An I/O error occurred while loading the variables: " + ExceptionUtils.toString(ioEx));
+
+			if (ioException != null) {
+				Skript.error("An I/O error occurred while loading the variables: " + ExceptionUtils.toString(ioException));
 				Skript.error("This means that some to all variables could not be loaded!");
 			}
+
 			try {
 				if (update2_1) {
 					Skript.info("[2.1] updating " + file.getName() + " to the new format...");
 				}
-				final File bu = FileUtils.backup(file);
-				Skript.info("Created a backup of " + file.getName() + " as " + bu.getName());
+
+				// Back up CSV file
+				File backupFile = FileUtils.backup(file);
+				Skript.info("Created a backup of " + file.getName() + " as " + backupFile.getName());
+
 				loadError = false;
-			} catch (final IOException ex) {
+			} catch (IOException ex) {
 				Skript.error("Could not backup " + file.getName() + ": " + ex.getMessage());
 			}
 		}
-		
+
 		if (update2_1) {
+			// Save variables in new format
 			saveVariables(false);
 			Skript.info(file.getName() + " successfully updated.");
 		}
-		
+
 		connect();
-		
-		saveTask = new Task(Skript.getInstance(), 5 * 60 * 20, 5 * 60 * 20, true) {
+
+		// Start the save task
+		saveTask = new Task(Skript.getInstance(), SAVE_TASK_DELAY, SAVE_TASK_PERIOD, true) {
 			@Override
 			public void run() {
 				if (changes.get() >= REQUIRED_CHANGES_FOR_RESAVE) {
```

Comment:
Shouldn't this check changes and then immediately set to 0? With `AtomicInteger#updateAndGet` perhaps?

Reply:
Then it'd have to be something like this:
![image](https://user-images.githubusercontent.com/29547183/215800518-cbd6418e-cea1-4667-abd9-a58039fdc5d4.png)
and I don't think it's better, cause of the double comparison.

Although the current scenario may indeed lose an increment due to threading, that's not such a bad thing. The only thing that happens to the atomic integer outside of the save thread is single increments, so worst case there is one (or even rarer a few) increments between the get and the set call, but that just means that for this `saveVariables` call and the next there will be a few more / less variables to save. The `saveVariables` function doesn't rely on the amount of changed variables to be exactly equal to some amount, so it should be fine.

However, I am adding a comment clarifying why it isn't needed.